site/docs/v1/tech/tutorials/integrate-flutter.md

```diff
@@ -0,0 +1,747 @@
+
+> This best current practice requires that native apps MUST NOT use embedded user-agents to perform authorization requests and allows that authorization endpoints MAY take steps to detect and block authorization requests in embedded user-agents.
+
+Moving forward, here's `loginAction`:
+
+```dart
+Future<void> loginAction() async {
+  setState(() {
+    isBusy = true;
+    errorMessage = '';
+  });
+
+  try {
+    final AuthorizationTokenResponse result =
+        await appAuth.authorizeAndExchangeCode(
+      AuthorizationTokenRequest(
+        FUSION_AUTH_CLIENT_ID,
+        FUSION_AUTH_REDIRECT_URI,
+        issuer: 'https://$FUSION_AUTH_DOMAIN',
+        scopes: <String>['openid'],
+      ),
+    );
+    log('data: $result');
+    final Map<String, Object> profile =
+        await getUserDetails(result.accessToken);
+
+    debugPrint('response: $profile');
+    await secureStorage.write(
+          key: 'refresh_token', value: result.refreshToken);
+    var gravatar = Gravatar(profile['email']);
+    var url = gravatar.imageUrl(
+      size: 100,
+      defaultImage: GravatarImage.retro,
+      rating: GravatarRating.pg,
+      fileExtension: true,
+    );
+    setState(() {
+      isBusy = false;
+      isLoggedIn = true;
+      name = profile['given_name'];
+      picture = url;
+    });
+  } on Exception catch (e, s) {
+    debugPrint('login error: $e - stack: $s');
+
+    setState(() {
+      isBusy = false;
+      isLoggedIn = false;
+      errorMessage = e.toString();
+    });
+  }
+}
+// ...
+```
+
+This calls `getUserDetails`, which takes the access token as a parameter and makes the HTTP GET request to the auth server. If successful, it will receive the user details in response. This is a standard OIDC call. This function is called inside the `initAction` function as well.
+
+```dart
+// ...
+Future<Map<String, Object>> getUserDetails(String accessToken) async {
+  const String url = 'https://$FUSION_AUTH_DOMAIN/oauth2/userinfo';
+  final http.Response response = await http.get(
+    url,
+    headers: <String, String>{'Authorization': 'Bearer $accessToken'},
+  );
+
+  if (response.statusCode == 200) {
+    return jsonDecode(response.body);
+  } else {
+    throw Exception('Failed to get user details');
+  }
+}
+// ...
+```
+
+To handle logging the user out, the `logoutAction` will be responsible for two things:
+
+* Remove the access token from secure storage
+* Set the state back to initial values
+
+These are all shown in the code snippet below:
+
+```dart
+Future<void> logoutAction() async {
+  await secureStorage.delete(key: 'refresh_token');
+  setState(() {
+    isLoggedIn = false;
+    isBusy = false;
+  });
+}
+```
+
+After editing `main.dart`, start up your emulators or real devices again.
+
+```shell
+flutter run -d all
+```
+
+You should see something similar to this demo:
+
+{% include _youtube-video.liquid youtubeid="EklSdaT0xfw" %}
+
```

Comment:
Let's add something at the end like the ruby on rails tutorial has.

Reply:
What do you mean?